File: src/jumpingOnClouds.js

class Node {
  constructor(data, next) {
   this.data = data;
   this.next = next;
  }

  set_next(node) {
    this.next = node;
  }
}

class List {
  constructor() {
    this.head = null;
  }

  add(data) {
    if (!this.head) {
      this.head = new Node(data);
    }
    else {
      const last_node = this.tail();
      last_node.set_next(new Node(data));
    }
  }

  tail() {
    let current_node = this.head;
    while (current_node.next) {
      current_node = current_node.next;
    }
    return current_node;
  }
}


const jumpingOnClouds = arr => {
  const list = new List();
  arr.forEach(item => list.add(parseInt(item)));
  let jumps = 0;
  let current_node = list.head;
  while (current_node.next) {
    if (current_node.next && current_node.next.next && current_node.next.next.data === 0) {
      jumps++;
      current_node = current_node.next.next;
    }
    else {
      jumps++;
      current_node = current_node.next;
    }
  }
  return jumps;
};

module.exports = jumpingOnClouds;

File: lib/jumping_on_clouds.rb
class CloudNode
  def initialize(data)
    @data = data
    @next = nil
  end

  def set_next(node)
    @next = node
  end

  def next
    @next
  end

  def data
    @data
  end
end

class CloudList
  def initialize(items = [])
    @head = nil
    items.each do |item|
     add item
    end
  end

  def head
    @head
  end

  def add(data)
    if @head.nil?
      @head = CloudNode.new data
    else
      get_tail.set_next CloudNode.new(data)
    end
  end

  private
    def get_tail
      current_node = @head
      while current_node.next do
        current_node = current_node.next
      end
      current_node
    end
end

def jumpingOnClouds(arr)
  list = CloudList.new arr
  current_node = list.head
  jumps = 0
  while current_node.next
    if current_node.next and current_node.next.next and current_node.next.next.data == '0'
      current_node = current_node.next.next
    else
      current_node = current_node.next
    end
    jumps = jumps + 1
  end
  jumps
end
